### PR TITLE
Implement python scalar conversions via ATen; allow localScalar if numel == 1

### PR DESCRIPTION
**Pull request by gchanan**
This implements the following:
[ATen]
1) Loosens the requirement on localScalar() that the tensor is a Scalar; now it only needs to be one element; Scalar(...) still has the old requirement because it makes the code simpler (and people probably don't call that constructor directly anymore because we provide nicer conversion functions).
2) Implement toCDouble, toCFloat, etc. so that it calls through localScalar() and gets the 1-element property; this is to match the pytorch scalar conversion API and https://github.com/pytorch/pytorch/issues/3839.
3) Add a is_nonzero native function which is functionally equivalent to toCBool, which doesn't exist because we don't have bool tensors (it adds a nicer error message as well).

[pytorch]
Uses the above ATen changes to implement Variable scalar conversions via ATen; there is a bit of complication around conversion from double to (python) int, because we want to avoid the ATen code path in that case because PyLongs don't overflow.  Also added a bunch of tests.

### CODE REVIEW

**Comment by gchanan** [comment]
CC @zdevito for the ATen changes.

File: tools/autograd/templates/python_variable_methods.cpp [hunk 1]
```diff
@@ -206,6 +206,40 @@ static PyObject * THPVariable_detach_(PyObject* self, PyObject* args)
...
+
+static PyObject * THPVariable_integral_scalar(PyObject* self, PyObject* args) {
+  HANDLE_TH_ERRORS
+  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
+  if (isFloatingType(self_.type().scalarType())) {
+    // we can't dispatch to toCLong here because we want to avoid ATen overflow checks;
+    // the python integral type (long in python2) can't overflow.
+    AutoGPU auto_gpu(self_);
+    Scalar localScalar = self_.get()->localScalar();
+    return THPUtils_packDoubleAsInt(localScalar.toDouble());
```
Comment (colesbury) [inline tools/autograd/templates/python_variable_methods.cpp]: These three lines could be:

```
return THPUtils_packDoubleAsInt(dispatch_to_CDouble(self_));
```

This has the advantage of releasing the GIL before doing a potentially blocking op.

**Review by colesbury — APPROVED** [review]
lgtm

**Comment by ezyang** [comment]
@gchanan Do you want this squashed?

**Comment by gchanan** [comment]
I'm going to merge this now; @zdevito let me know if you want any changes and we can do them in subsequent PRs.

**Pull request merged**